src/Serialization/main.jl

```diff
@@ -1,9 +1,135 @@
+# save_object / load_object
+
+# These functions are at the core of the serialization and are the first functions
+# that should be implemented when working on the serialization of a new type.
+# Here is where one should use functions save_data_dict and save_data_array to structure
+# the serialization. The examples show they can be used to save data using the structure
+# of an array or dict. Each nested call to save_data_dict or save_data_array should be
+# called with a key that can be passed as the second parameter.
+
+#  Examples
+#  function save_object(s::SerializerState, obj::NewType)
+#    save_data_array(s) do
+#      save_object(s, obj.1)
+#      save_object(s, obj.2)
+#
+#      save_data_dict(s) do
+#        save_object(s, obj.3, :key1)
+#        save_object(s, obj.4, :key2)
+#      end
+#    end
+#  end
+#
+#  This will result in a data format that looks like
+#  [
+#    obj.1,
+#    obj.2,
+#    {
+#      "key1": obj.3,
+#      "key2": obj.4
+#    }
+#  ]
+
+#  function save_object(s::SerializerState, obj::NewType)
+#    save_data_dict(s) do
+#      save_object(s, obj.1, :key1)
+#      save_data_array(s, :key2) do
+#        save_object(s, obj.3)
+#        save_typed_object(s, obj.4) # This is ok
+#      end
+#    end
+#  end
+
+#  This will result in a data format that looks like
+#  {
+#    "key1": obj.1,
+#    "key2":[
+#      obj.3,
+#      {
+#        "type": "Type of obj.4",
+#        "data": obj.4
+#      }
+#    ]
+#  }
+#
+# This is ok
+# function save_object(s::SerializerState, obj:NewType)
+#   save_object(s, obj.1)
+# end
+
+# This will throw an error 
+# function save_object(s::SerializerState, obj:NewType)
+#   save_object(s, obj.1, :key)
+# end
+
+# If you insist on having a key you should, first open a save_data_dict.
+
+# function save_object(s::SerializerState, obj:NewType)
+#   save_data_dict(s) do
+#     save_object(s, obj.1, :key)
+#   end
+# end
+#
+################################################################################
+# save_type_params / load_type_params
+
+# The serialization mechanism stores data in the format of a tree, with the
+# exception that some nodes may point to a shared reference. The "data branch"
+# is anything that is a child node of a data node, whereas the "type branch" is
+# any information that is stored in a node that is a child of a type node.
+# Avoiding type information inside the data branch will lead to a more
+# efficient serialization format. When the uses_params is set while calling
+# @register_serialization_type (de)serialization will use 
+# save_type_params / load_type_params to format the type information.
+# In general we expect that implementing a save_type_params and load_type_params
+# should not always be necessary. Many types will serialize their types
+# in a similar fashion for example serialization of a FieldElem will
+# use the save_type_params / load_type_params from RingElem since in both cases
+# the only parameter needed for such types is their parent.
```

Comment:
OK, I still have questions. But it's fine by me to merge this already, and improve it later.